### DISCUSSION TITLE
[Question] How to give type to Zod Arrays

### DISCUSSION THREAD

**Discussion by SaadAhmadSaddiqui** [colinhacks/zod#1754] [Q&A] [answered]
So I have a type 

e.g. `type QualityTypes = 'thing1' | 'thing2' |  'thing3' |  'thing4' |  'thing5'`

Now I want a zod array with `QualityTypes` as the type of the array when I infer the value of the zod object.

e.g.

```typescript
const personSchema = z.object({
    name: z.string().required(),
    qualities: z.array(z.string()).optional()
})

type Person = z.infer<typeof personSchema>;

/*
Person = {
    name: string;
    qualities: string[]
}
*/
```  

I wanna make it so that the qualities property says, `qualities: QualityTypes[]`

**Comment by daniellwdb** [comment]
You can use [brand](https://github.com/colinhacks/zod#brand) for this like: `z.array(z.string().brand<QualityTypes>()).optional()`. This will give you a branded type when inferring the schema type and you can either take the easy route and accept `z.BRAND<QualityTypes>[]` instead of `QualityTypes[]`  where you intend on passing it to make the returned branded type structurally compatible. Or you can create a recursive conditional mapped type to strip away the brands and leave you with `qualities: QualityTypes[]` (or multiple single-purpose utility types that remove the brands)

**Comment by JacobWeisenburger** [comment, marked as answer]
This is how I would solve this. But it doesn't have `qualities: QualityTypes[]` in the type like you wanted. However it does ensure that the values are valid QualityTypes. 

```ts
const personSchema = z.object( {
    qualities: z.enum( [ 'thing1', 'thing2', 'thing3' ] ).array().optional(),
} )

type Person = z.infer<typeof personSchema>
// type Person = {
//     qualities?: ( "thing1" | "thing2" | "thing3" )[] | undefined
// }

console.log(
    personSchema.safeParse( { qualities: [ 'thing1', 'thing2' ] } ).success // true
)
console.log(
    personSchema.safeParse( { qualities: [ 'foo', 'bar' ] } ).success // false
)
```

---

This way gets you closer to what you are asking for (`qualities: QualityTypes[]`), but it doesn't ensure that the values are valid QualityTypes.
```ts
const personSchema = z.object( {
    qualities: z.string().brand<'QualityTypes'>().array().optional(),
} )

type Person = z.infer<typeof personSchema>
// type Person = {
//     qualities?: ( string & z.BRAND<"QualityTypes"> )[] | undefined
// }

console.log(
    personSchema.safeParse( { qualities: [ 'thing1', 'thing2' ] } ).success // true
)
console.log(
    personSchema.safeParse( { qualities: [ 'foo', 'bar' ] } ).success // true
    // I imagine you would want this to be false
)
```

I'm not sure that there is a good way to make the type show `qualities: QualityTypes[]`.
